p2p/host/relay/autorelay.go

```diff
@@ -253,6 +318,54 @@ func (ar *AutoRelay) doUpdateAddrs() {
 	ar.addrs = raddrs
 }
 
+// This function cleans up a relay's address set to remove private addresses and curtail
+// addrsplosion. For the latter, we use the following heuristic:
+// - if the address set includes a (tcp) address with the default port 4001,
+//   we remove all tcp addrs with a different port
+func cleanupAddressSet(pi pstore.PeerInfo) pstore.PeerInfo {
+	// pass-1: find default port
+	has4001 := false
+	for _, addr := range pi.Addrs {
+		port, err := tcpPort(addr)
+		if err != nil {
+			continue
+		}
+		if port == 4001 {
+			has4001 = true
+			break
+		}
+	}
+
+	// pass-2: cleanup
+	var newAddrs []ma.Multiaddr
+
+	for _, addr := range pi.Addrs {
+		if manet.IsPrivateAddr(addr) {
+			continue
+		}
+
+		if has4001 {
+			port, err := tcpPort(addr)
+			if err == nil && port != 4001 {
+				continue
+			}
+		}
+
+		newAddrs = append(newAddrs, addr)
+	}
+
+	return pstore.PeerInfo{ID: pi.ID, Addrs: newAddrs}
+}
+
+func tcpPort(addr ma.Multiaddr) (int, error) {
+	val, err := addr.ValueForProtocol(ma.P_TCP)
```

Comment:
This function should really just be avoided. I know it's annoying but it's much better to actually parse the multiaddr.

Reply:
ok, should I parse by component and look for tcp?

Reply:
I killed the function altogether.

Reply:
I introduced `ma.ForEach` and friends to try to make this a bit easier. It's still _extremely_ painful and I'd like to try some kind of pattern matching solution but that's a bit tricky in go (no macros :cry:).

Reply:
Yeah, the new parsing code uses `ma.ForEach`.